src/operator/tensor/broadcast_reduce_op.h

```diff
@@ -1103,21 +1139,44 @@ inline void BroadcastComputeImpl(const nnvm::NodeAttrs& attrs,
           out_shape[i] = 1;
         }
       }
+      struct shape_and_stride aux_data;
+      index_t iter = dst_shape.ndim() - 1;
+      aux_data.out_stride[iter] = 1;
+      aux_data.in_stride[iter] = 1;
+      aux_data.input_shape[iter] = in_shape[iter];
+      aux_data.output_shape[iter] = out_shape[iter];
+      iter--;
+      for (; iter >= 0; --iter) {
+        aux_data.out_stride[iter] = aux_data.out_stride[iter+1] * out_shape[iter+1];
+        aux_data.in_stride[iter] = aux_data.in_stride[iter+1] * in_shape[iter+1];
+        aux_data.input_shape[iter] = in_shape[iter];
+        aux_data.output_shape[iter] = out_shape[iter];
+      }
       if (dst_shape.ndim() == 2) {
         Tensor<xpu, 2, OType> out =
           outputs[0].get_with_shape<xpu, 2, OType>(dst_shape.get<2>(), s);
         Tensor<xpu, 2, IType> data =
           inputs[0].get_with_shape<xpu, 2, IType>(src_shape.get<2>(), s);
-        Kernel<broadcast_kernel<mshadow_op::identity>, xpu>::Launch(
-          s, out.shape_.Size(), data.dptr_, out.dptr_, in_shape, out_shape, req[0], 2);
+        if (ctx.run_ctx.get_ctx().dev_type == Context::kGPU) {
+          Kernel<broadcast_kernel_gpu<mshadow_op::identity>, xpu>::Launch(
```

Comment:
Maybe I missed something, but the gpu and cpu kernels look highly similar. Adding a branch here makes the code very fragmented since this entire function `template<typename xpu> inline void BroadcastComputeImpl` is supposed to be generic template for both CPU and GPU. 

I think we need a better way to code this. Two approaches I can think of
(1) consolidate cpu and gpu kernel
(2) instead of a generic template, break it into two different functions.

(1) is preferred in my opinion.